File: src/transpiler-output/FocusSequenceTracker.ts
import {
    ConfidenceEngineConfig,
    FocusEventData,
    KeydownEventData,
} from './confidence-types';

// Namespace: Minimact.Workers

export class FocusSequenceTracker {
    private focusSequence: string[];
    private currentFocusIndex: number;
    private lastTabPressTime: number;
    private config: ConfidenceEngineConfig;
    constructor(config: ConfidenceEngineConfig) {
        this.config = config;
    }

    trackFocus(eventData: FocusEventData): void {
        const elementId: string = eventData.elementId;
        const existingIndex: number = Array.indexOf(this.focusSequence, elementId);
        if (existingIndex != -1) {
            this.currentFocusIndex = existingIndex;
        } else {
            const newSequence: string[] = new Array<string>(this.focusSequence.length + 1);
            Array.copy(this.focusSequence, newSequence, this.focusSequence.length);
            newSequence[this.focusSequence.length] = elementId;
            this.focusSequence = newSequence;
            this.currentFocusIndex = this.focusSequence.length - 1;
        }
    }

    trackKeydown(eventData: KeydownEventData): void {
        if (eventData.key == "Tab") {
            this.lastTabPressTime = eventData.timestamp;
        }
    }

    registerFocusSequence(elementIds: string[]): void {
        this.focusSequence = elementIds;
    }

    export class FocusConfidenceResult {
        confidence: number;
        leadTime: number;
        reason: string;
    }

    calculateFocusConfidence(elementId: string, currentTime: number): FocusConfidenceResult {
        const timeSinceTab: number = currentTime - this.lastTabPressTime;
        if (timeSinceTab > 100) {
            return { confidence: 0, leadTime: 0, reason: "no recent Tab press" };
        }
        if (this.focusSequence.length == 0 || this.currentFocusIndex == -1) {
            return { confidence: 0, leadTime: 0, reason: "no focus sequence data" };
        }
        const nextIndex: number = (this.currentFocusIndex + 1) % this.focusSequence.length;
        const nextElementId: string = this.focusSequence[nextIndex];
        if (nextElementId == elementId) {
            return { confidence: this.config.focusHighConfidence, leadTime: 50, reason: `Tab pressed, next in sequence (index ${nextIndex})` };
        }
        return { confidence: 0, leadTime: 0, reason: "not next in sequence" };
    }

    export class FocusPredictionResult {
        elementId: string;
        confidence: number;
        leadTime: number;
        reason: string;
    }

    predictNextFocus(): FocusPredictionResult {
        if (this.focusSequence.length == 0 || this.currentFocusIndex == -1) {
            return { elementId: null, confidence: 0, leadTime: 0, reason: "no focus sequence" };
        }
        const nextIndex: number = (this.currentFocusIndex + 1) % this.focusSequence.length;
        const nextElementId: string = this.focusSequence[nextIndex];
        return { elementId: nextElementId, confidence: this.config.focusHighConfidence, leadTime: 50, reason: `Tab navigation to index ${nextIndex}` };
    }

    clear(): void {
        this.focusSequence = new Array<string>(0);
        this.currentFocusIndex = -1;
        this.lastTabPressTime = 0;
    }

}

File: src/Minimact.Workers/FocusSequenceTracker.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Minimact.Workers
{
    /// <summary>
    /// Focus Sequence Tracker
    ///
    /// Tracks focus events and Tab key presses to predict the next focused element
    /// with very high confidence (Tab sequence is deterministic).
    /// </summary>
    public class FocusSequenceTracker
    {
        private string[] focusSequence = new string[0]; // Ordered list of focusable element IDs
        private int currentFocusIndex = -1;
        private double lastTabPressTime = 0;
        private ConfidenceEngineConfig config;

        public FocusSequenceTracker(ConfidenceEngineConfig config)
        {
            this.config = config;
        }

        /// <summary>
        /// Record a focus event
        /// </summary>
        public void TrackFocus(FocusEventData eventData)
        {
            string elementId = eventData.ElementId;

            // Update focus sequence
            int existingIndex = Array.IndexOf(this.focusSequence, elementId);
            if (existingIndex != -1)
            {
                this.currentFocusIndex = existingIndex;
            }
            else
            {
                // New element in sequence
                string[] newSequence = new string[this.focusSequence.Length + 1];
                Array.Copy(this.focusSequence, newSequence, this.focusSequence.Length);
                newSequence[this.focusSequence.Length] = elementId;
                this.focusSequence = newSequence;
                this.currentFocusIndex = this.focusSequence.Length - 1;
            }
        }

        /// <summary>
        /// Record a keydown event
        /// </summary>
        public void TrackKeydown(KeydownEventData eventData)
        {
            if (eventData.Key == "Tab")
            {
                this.lastTabPressTime = eventData.Timestamp;
            }
        }

        /// <summary>
        /// Register the focus sequence for elements
        /// (Called when elements are registered with the worker)
        /// </summary>
        public void RegisterFocusSequence(string[] elementIds)
        {
            this.focusSequence = elementIds;
        }

        /// <summary>
        /// Focus confidence result
        /// </summary>
        public class FocusConfidenceResult
        {
            public double Confidence { get; set; }
            public double LeadTime { get; set; }
            public string Reason { get; set; }
        }

        /// <summary>
        /// Calculate focus confidence for an element
        ///
        /// Returns confidence [0-1] that element will receive focus
        /// </summary>
        public FocusConfidenceResult CalculateFocusConfidence(string elementId, double currentTime)
        {
            // Was Tab pressed recently? (within 100ms)
            double timeSinceTab = currentTime - this.lastTabPressTime;
            if (timeSinceTab > 100)
            {
                return new FocusConfidenceResult
                {
                    Confidence = 0,
                    LeadTime = 0,
                    Reason = "no recent Tab press"
                };
            }

            // Do we know the focus sequence?
            if (this.focusSequence.Length == 0 || this.currentFocusIndex == -1)
            {
                return new FocusConfidenceResult
                {
                    Confidence = 0,
                    LeadTime = 0,
                    Reason = "no focus sequence data"
                };
            }

            // Calculate next focus index (forward)
            int nextIndex = (this.currentFocusIndex + 1) % this.focusSequence.Length;
            string nextElementId = this.focusSequence[nextIndex];

            // Is this element the next in sequence?
            if (nextElementId == elementId)
            {
                // Very high confidence - Tab sequence is deterministic
                return new FocusConfidenceResult
                {
                    Confidence = this.config.FocusHighConfidence, // 0.95
                    LeadTime = 50, // Very short lead time (~50ms for focus to occur)
                    Reason = $"Tab pressed, next in sequence (index {nextIndex})"
                };
            }

            // Check if Shift+Tab (backward navigation)
            // For now, assume forward Tab only
            // TODO: Track Shift key for backward navigation

            return new FocusConfidenceResult
            {
                Confidence = 0,
                LeadTime = 0,
                Reason = "not next in sequence"
            };
        }

        /// <summary>
        /// Focus prediction result
        /// </summary>
        public class FocusPredictionResult
        {
            public string ElementId { get; set; }
            public double Confidence { get; set; }
            public double LeadTime { get; set; }
            public string Reason { get; set; }
        }

        /// <summary>
        /// Calculate focus confidence when Tab is pressed
        /// (More proactive - predicts immediately on Tab)
        /// </summary>
        public FocusPredictionResult PredictNextFocus()
        {
            if (this.focusSequence.Length == 0 || this.currentFocusIndex == -1)
            {
                return new FocusPredictionResult
                {
                    ElementId = null,
                    Confidence = 0,
                    LeadTime = 0,
                    Reason = "no focus sequence"
                };
            }

            int nextIndex = (this.currentFocusIndex + 1) % this.focusSequence.Length;
            string nextElementId = this.focusSequence[nextIndex];

            return new FocusPredictionResult
            {
                ElementId = nextElementId,
                Confidence = this.config.FocusHighConfidence, // 0.95
                LeadTime = 50,
                Reason = $"Tab navigation to index {nextIndex}"
            };
        }

        /// <summary>
        /// Clear history
        /// </summary>
        public void Clear()
        {
            this.focusSequence = new string[0];
            this.currentFocusIndex = -1;
            this.lastTabPressTime = 0;
        }
    }
}